galleon-plugins/src/main/java/org/wildfly/galleon/plugin/WfInstallPlugin.java

```diff
@@ -543,6 +564,64 @@ public void resolve(MavenArtifact artifact) throws ProvisioningException {
         }
     }
 
+    private List<MavenArtifact> listMavenArtifactsInModules() throws ProvisioningException {
+        List<MavenArtifact> allModuleArtifacts = new ArrayList<>();
+        for (Entry<Path, PackageRuntime> entry : jbossModules.entrySet()) {
+            final PackageRuntime pkg = entry.getValue();
+            try {
+                allModuleArtifacts.addAll(findArtifacts(pkg, entry.getKey()));
+            } catch (IOException e) {
+                throw new ProvisioningException("Failed to process JBoss module XML template for feature-pack "
+                                                   + pkg.getFeaturePackRuntime().getFPID() + " package " + pkg.getName(), e);
+            }
+        }
+        return allModuleArtifacts;
+    }
+
+    private List<MavenArtifact> findArtifacts(PackageRuntime pkg, Path moduleXmlRelativePath) throws ProvisioningException, IOException {
+        final Path moduleTemplateFile = pkg.getResource(WfConstants.PM, WfConstants.WILDFLY, WfConstants.MODULE).resolve(moduleXmlRelativePath);
+        final Path targetPath = runtime.getStagedDir().resolve(moduleXmlRelativePath.toString());
+        final Map<String, String> versionProps = fpArtifactVersions.get(pkg.getFeaturePackRuntime().getFPID().getProducer());
+        ModuleTemplate moduleTemplate = new ModuleTemplate(pkg, moduleTemplateFile, targetPath);
+        if (!moduleTemplate.isModule()) {
+            return Collections.emptyList();
+        }
+
+        final Elements artifacts = moduleTemplate.getArtifacts();
+        if (artifacts == null) {
+            return Collections.emptyList();
+        }
+
+        List<MavenArtifact> res = new ArrayList<>();
+        final int artifactCount = artifacts.size();
+        for (int i = 0; i < artifactCount; i++) {
+            final AbstractModuleTemplateProcessor.ModuleArtifact moduleArtifact = new AbstractModuleTemplateProcessor.ModuleArtifact(artifacts.get(i), versionProps, log, artifactInstaller, channelArtifactResolution);
+            final MavenArtifact mavenArtifact = moduleArtifact.getUnresolvedArtifact();
+            if (mavenArtifact != null) {
+                res.add(mavenArtifact);
+            }
+        }
+
+        return res;
+    }
+
+    private Map<MavenArtifact, MavenArtifact> resolveAll(List<MavenArtifact> allModuleArtifacts) throws MavenUniverseException {
+        Map<MavenArtifact, MavenArtifact> artifactCache = new HashMap<>();
+        for (MavenArtifact a : allModuleArtifacts) {
+            final MavenArtifact b = new MavenArtifact();
+            b.setGroupId(a.getGroupId());
+            b.setArtifactId(a.getArtifactId());
+            b.setExtension(a.getExtension());
+            b.setClassifier(a.getClassifier());
+            b.setVersion(a.getVersion());
+            b.setVersionRange(a.getVersionRange());
+            artifactCache.put(b, a);
+        }
+        maven.resolveAll(allModuleArtifacts);
```

Comment:
If we have a pre-populated cache, should be: maven.resolveAll(artifactCache.keySet());
This implies that the maven.resolveAll method signature be changed from List<MavenArtifact> to Collection<MavenArtifact>.

Reply:
that would put the resolved artifacts as the keys. I think that would prevent them from matching when they're resolved in ModuleProcessors.

Reply:
nevermind, didn't click that it's related to earlier comment. I'll take a look